### DISCUSSION TITLE
🧪 Return stats for numerical facets like `min`, `max`

### DISCUSSION THREAD

**Discussion by 3kiran** [meilisearch/product#117] [Feedback & Feature Proposal]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by Kerollmops** [comment]
Hey @3kiran,

We are working on the new version of the search engine. We will be able to and will need to expose the minimum and maximum values for a facet number for [the instant-meilisearch extension](https://github.com/meilisearch/instant-meilisearch) to work correctly with facet ranges. I want to keep this issue open to close it when we are done with this feature and notify you :)

Thank you, and have a good day!

**Comment by curquiza** [comment]
@gmourier, this one can interest you

**Comment by gmourier** [comment]
Cross-linking a need expressed [here](https://github.com/meilisearch/instant-meilisearch/issues/926)

The idea of a workaround on the instant-meilisearch side is discussed. @bidoubiwa points out the potential limits to this workaround.

**Comment by gmourier** [comment]
Hello everyone,

EDIT: [It's now available in the v1.1 release candidate](https://github.com/meilisearch/product/discussions/117#discussioncomment-5238276)

We just released a prototype that allows you to get the `min` / `max` statistics for a facet to ease the implementation of a range slider component.

When using the `facets` parameter at search, the facets with numeric values are displayed in a new section of the search response called `facetStats` that contains, per facet, the numeric `min` and `max` values of the hits returned by the search.

Note: It ignores string values even if parseable. e.g `"21"` won't be considered by the engine when computing the `facetStats` `min` and `max`. 

## How to get the prototype?

Using docker, use the following command: `docker run -p 7700:7700 -v $(pwd)/meili_data:/meili_data getmeili/meilisearch:prototype-facet-stats-0`

From source, compile Meilisearch on the `prototype-facet-stats-0` tag

## How to use the prototype?

⚠️ Do not use the prototype in production ⚠️

### Usage Example

```bash
curl \
  -X POST 'http://localhost:7700/indexes/:indexUid/search' \
  -H 'Content-Type: application/json'
```

#### Request Payload

```json
{ 
    "q": "LL6", 
    "facets": ["mass", "recclass"], 
    "limit": 5 
}
```

#### Response Payload

```json
{
  "hits": [
    {
      "name": "Niger (LL6)",
      "id": "16975",
      "nametype": "Valid",
      "recclass": "LL6",
      "mass": 3.3,
      "fall": "Fell"
    },
    {
      "name": "Appley Bridge",
      "id": "2318",
      "nametype": "Valid",
      "recclass": "LL6",
      "mass": 15000,
      "fall": "Fell",
      "_geo": {
        "lat": 53.58333,
        "lng": -2.71667
      }
    },
    {
      "name": "Athens",
      "id": "4885",
      "nametype": "Valid",
      "recclass": "LL6",
      "mass": 265,
      "fall": "Fell",
      "_geo": {
        "lat": 34.75,
        "lng": -87.0
      }
    },
    {
      "name": "Bandong",
      "id": "4935",
      "nametype": "Valid",
      "recclass": "LL6",
      "mass": 11500,
      "fall": "Fell",
      "_geo": {
        "lat": -6.91667,
        "lng": 107.6
      }
    },
    {
      "name": "Benguerir",
      "id": "30443",
      "nametype": "Valid",
      "recclass": "LL6",
      "mass": 25000,
      "fall": "Fell",
      "_geo": {
        "lat": 32.25,
        "lng": -8.15
      }
    }
  ],
  "query": "LL6",
  "processingTimeMs": 15,
  "limit": 5,
  "offset": 0,
  "estimatedTotalHits": 42,
  "facetDistribution": {
    "mass": {
      "110000": 1,
      "11500": 1,
      "1161": 1,
      "12000": 1,
      "1215.5": 1,
      "127000": 1,
      "15000": 1,
      "1676": 1,
      "1700": 1,
      "1710.5": 1,
      "18000": 1,
      "19000": 1,
      "220000": 1,
      "2220": 1,
      "22300": 1,
      "25000": 2,
      "265": 1,
      "271000": 1,
      "2840": 1,
      "3.3": 1,
      "3000": 1,
      "303": 1,
      "32000": 1,
      "34000": 1,
      "36.1": 1,
      "45000": 1,
      "460": 1,
      "478": 1,
      "483": 1,
      "5500": 2,
      "600": 1,
      "6000": 1,
      "67.8": 1,
      "678": 1,
      "680.5": 1,
      "6930": 1,
      "8": 1,
      "8300": 1,
      "840": 1,
      "8400": 1
    },
    "recclass": {
      "L/LL6": 3,
      "LL6": 39
    }
  },
  "facetStats": {
    "mass": {
      "min": 3.3,
      "max": 271000.0
    }
  }
}
```

Thanks to anyone taking the time to test this prototype and sharing us feedback 🙇‍♂️

If it's more convenient for you, we have a [dedicated discord discussion](https://discord.com/channels/1006923006964154428/1075349145067466752/1075349145067466752). Feel free to join! ✨

**Reply by bidoubiwa** [reply]
While, I think the design choice is the easy choice as it follows an already existing API. However, should `facetStats` and `facetDistribution` be two different objects? Or should `distribution` be part of `facetStats`. 
```
facetStats: {
   mass: {
      min: 3,
      max: 212739,
     distribution: {
       8": 1,
      "8300": 1,
      "840": 1,
      "8400": 1
     // etc
   }
}
}
```
I know it would introduce a breaking change, so if we agree it would be for v2.

Meanwhile, I think `facetStats` is a good option.

**Reply by Fronix** [reply]
I feel `facetStats` and `facetDistribution` should be separate. Maybe it could be called `facetRange` instead? Since (I'm assuming) it will only have a min and max values for facets.

Not really sure what `facetStats` even means in this context, unless it's made to serve multiple purposes.

**Reply by gmourier** [reply]
Thank you! We may want to add `avg` to it in the future @Fronix

**Comment by nmtri-dev** [comment]
I tested the prototype, so far it works for me. I just had some questions:
- Will we have option to disable facetStats/ facetDistribution for specific attribute? Sometimes, I want to get only facetStats and vice versa.
- Do we have a timeline when this feature will be released?

**Reply by dureuill** [reply]
Hey @nmtri-dev :wave:

- Can you tell me more about the use case for disabling facetStats / facetDistribution?
- The feature will be released as part of v1.1. Release candidates are set to appear at the beginning of March, and the final v1.1 at the beginning of April.

Thank you for testing the prototype and taking the time of giving us feedback :heart:

**Reply by nmtri-dev** [reply]
So my case is I want to aggregate products' prices. The things I need is min and max of prices, the list of 100 prices (or maybe longer if I change `maxValuesPerFacet`) returned along would be redundant.

**Reply by dureuill** [reply]
Thank you for answering. I understand that returning the entire distribution would be redundant if all you need is the min/max values, I'm trying to understand how much of an issue the redundancy causes.

Is your concern maybe performance-related? Or is it rather related to the human readability of the response?

**Reply by nmtri-dev** [reply]
Just performance-related concern for me.

**Comment by gmourier** [comment]
Hello, everyone following this discussion 👋 

We have just released the first RC (release candidate) of Meilisearch containing this new implementation!

You can test it by using
-  the [release assets](https://github.com/meilisearch/meilisearch/releases)
- the [Meilisearch Docker image](https://hub.docker.com/repository/docker/getmeili/meilisearch/general)

```bash
docker run -it --rm -p 7700:7700 -v $(pwd)/meili_data:/meili_data getmeili/meilisearch:v1.1.0-rc.0
```

You are welcome to communicate any feedback about this new implementation in this discussion.

If you encounter any bugs, please report them [here](https://github.com/meilisearch/meilisearch/issues/new?assignees=&labels=&template=bug_report.md&title=).
Thanks in advance for your help and your involvement in Meilisearch ❤️ 

🎉 Official and stable release containing this change will be available on 3rd April 2023

⚠️ RC (release candidates) are not recommended for production

**Comment by gmourier** [comment]
Hey folks 👋

[v1.1](https://github.com/meilisearch/meilisearch/releases/tag/v1.1.0) has been released and features the `facetStats` object when search results are returned during a faceted search. ✨

📚https://docs.meilisearch.com/learn/advanced/faceted_search.html#facet-stats
